File: src/Day14.kt


  import java.io.File
  import java.io.FileNotFoundException
  import java.nio.charset.Charset

  object Day14 {
  data class Problem14(val polymerFormula : String, val pairs : Map<String, String>)

  fun parseInput(s : String) : Problem14 {
    val polymerFormula = s.split("\n").first()
    var pairRegex = "([A-Za-z]+) -> ([A-Za-z]+)".toRegex()
    val pairs = s.split("\n")
      .drop(2)
      .mapNotNull { pairRegex.matchEntire(it) }
      .map { match -> match.groupValues.toList() }
      .associate { Pair(it[1], it[2]) }
    return Problem14(polymerFormula, pairs)
  }

  fun evaluatePolymer(problem14 : Problem14, steps : Int) : Long {
    var polymerFormula : String = problem14.polymerFormula
    for (i in 0 until steps) {
      var newPolymerFormula : String = ""
      var lastChars = ""
      for (c in polymerFormula) {
        lastChars += c
        if (lastChars.length > 2) {
          lastChars = lastChars.substring(1)
        }
        val replacement = problem14.pairs.getOrDefault(lastChars, "")
        newPolymerFormula += replacement + c
      }
      polymerFormula = newPolymerFormula
    }
    val test = polymerFormula.map { it }
      .groupBy{ it }
      .mapValues { it.value.size.toLong() }
    return test.values.maxOf { it } - test.values.minOf { it }
  }

  fun evaluatePolymerSmart(problem14 : Problem14, steps : Int) : Long {
    var polymerFormula : String = problem14.polymerFormula
    // for (i in 0 until steps) {
    var frequencyMap = mutableMapOf<String, Long>()
    var partialFormula = ""
    for (c in polymerFormula.drop(1).dropLast(1)) {
      partialFormula += c
      if (partialFormula.length >= 2) {
        frequencyMap[partialFormula.takeLast(2)] = 1
      }
    }

    var first = polymerFormula.take(2)
    var last = polymerFormula.takeLast(2)

    for (i in 0 until steps) {
      var newFrequencyMap = mutableMapOf<String, Long>()
      for (key in frequencyMap.keys) {
        val replacement = problem14.pairs.getOrDefault(key, "")
        if (replacement == "") {
          continue
        }
        val leftReplacement = key[0] + replacement
        val rightReplacement = replacement + key[1]
        newFrequencyMap[leftReplacement] = newFrequencyMap.getOrDefault(leftReplacement, 0) + frequencyMap.getOrDefault(key, 0)
        newFrequencyMap[rightReplacement] = newFrequencyMap.getOrDefault(rightReplacement, 0) + frequencyMap.getOrDefault(key, 0)
      }

      val lastReplacement = problem14.pairs.getOrDefault(last, "")
      val lastLeftReplacement = last[0] + lastReplacement
      newFrequencyMap[lastLeftReplacement] = 1 + newFrequencyMap.getOrDefault(lastLeftReplacement, 0)

      val firstReplacement = problem14.pairs.getOrDefault(first, "")
      val lastFirstReplacement = firstReplacement + first[1]
      newFrequencyMap[lastFirstReplacement] = 1 + newFrequencyMap.getOrDefault(lastFirstReplacement, 0)

      first = first[0] + firstReplacement
      last = lastReplacement + last[1]
      frequencyMap = newFrequencyMap
    }

    val wat = frequencyMap
      .map { listOf(Pair(it.key[0], it.value)) }
      .flatten()
      .groupBy { it.first }
      .mapValues { p -> p.value.sumOf { it.second } }
      .toMutableMap()
    wat[first[0]] = wat.getOrDefault(first[0], 0) + 1
    wat[last[0]] = wat.getOrDefault(last[0], 0) + 1
    wat[last[1]] = wat.getOrDefault(last[1], 0) + 1
    return wat.values.maxOf { it } - wat.values.minOf { it }
  }

    @JvmStatic
  fun main(args: Array<String>) {
    var fileName: String = args[0]
    try {
      val problem = parseInput( File("inputd14.txt").reader(Charset.forName("ASCII")).readText())
      val part1Result = evaluatePolymer(problem, 10)
      val part2Result = evaluatePolymerSmart(problem, 40)
      println("Day 14 part 1 result: $part1Result")
      println("Day 14 part 2 result: $part2Result")
    } catch (e: FileNotFoundException) {
      println("Input file does not exist ($fileName)")
    }
  }
}

File: src/Day14.scala
import scala.io.Source

object Day14 {

  def main(args: Array[String]) = {
    val filename = "inputd14.txt"

    val input=Source.fromFile(filename).getLines().toList

    var s = input(0)
    //CH -> B
    val pairs=input.slice(2,input.size).map(x =>{
      val t=x.split(" -> ")
     (t(0),t(1))
    }).toMap

    for (j <- 1 to 40)
    {
      var t = ""
      for (i <- 0 to s.length - 2) {
       // println(pairs.get(s.slice(i, i + 2)))
        val k=pairs.get(s.slice(i, i + 2))
        if (k.get!="") {
          t = t + s(i) + k.get
        } else {
          t = t + s(i)
        }
      }
      t = t + s(s.length-1)
      println(t.length)
      s = t
    }

    println( s.groupBy(c =>c ).map(p => p._2.length).max-  s.groupBy(c =>c ).map(p => p._2.length).min)
 //s.groupBy(c =>c )


   /// println(pairs)
  }
}

/**
 * After step 1: NCNBCHB
 *               NCNBCHB
After step 2: NBCCNBBBCBHCB
              NBCCNBBBCBHCB
After step 3: NBBBCNCCNBBNBNBBCHBHHBCHB
              NBBBCNCCNBBNBNBBCHBHHBCHB
After step 4: NBBNBNBBCCNBCNCCNBBNBBNBBBNBBNBBCBHCBHHNHCBBCBHCB
              NBBNBNBBCCNBCNCCNBBNBBNBBBNBBNBBCBHCBHHNHCBBCBHCB
 */
